pkg/ec2provider/ec2provider.go

```diff
@@ -0,0 +1,274 @@
+	p.requestQueue.requestQueueLock.RUnlock()
+	return length
+}
+
+func (p *ec2ProviderImpl) getPrivateDNSNameCache(id string) (string, error) {
+	p.privateDNSCache.dnsCacheLock.RLock()
+	defer p.privateDNSCache.dnsCacheLock.RUnlock()
+	name, ok := p.privateDNSCache.dnsCache[id]
+	if ok {
+		return name, nil
+	}
+	return "", errors.New("instance id not found")
+}
+
+// GetPrivateDNS looks up the private DNS from the EC2 API
+func (p *ec2ProviderImpl) GetPrivateDNSName(id string) (string, error) {
+	privateDNSName, err := p.getPrivateDNSNameCache(id)
+	if err == nil {
+		return privateDNSName, nil
+	}
+	logrus.Infof("Missed the cache for the InstanceId = %s Verifying if its already in requestQueue ", id)
+	// check if the request for instanceId already in queue.
+	if p.getRequestInQueueForInstanceId(id) {
+		logrus.Infof("Found the InstanceId:= %s request In Queue waiting in 5 seconds loop ", id)
+		for i := 0; i < total_iteration_for_wait_interval; i++ {
+			time.Sleep(default_wait_interval * time.Millisecond)
+			privateDNSName, err := p.getPrivateDNSNameCache(id)
+			if err == nil {
+				return privateDNSName, nil
+			}
+		}
+		return "", fmt.Errorf("failed to find node %s in PrivateDNSNameCache returning from loop", id)
+	}
+	logrus.Infof("Missed the requestQueue cache for the InstanceId = %s", id)
+	p.setRequestInQueueForInstanceId(id)
+	requestQueueLength := p.getRequestInQueueSize()
+	//The code verifies if the requestQuqueMap size is greater than max request in flight with rate
+	//limiting then writes to the channel where we are making batch ec2:DescribeInstances API call.
+	if requestQueueLength > max_allowed_single_request {
+		logrus.Infof("Writing to buffered channel for instance Id %s ", id)
+		p.instanceIdsChannel <- id
+		return p.GetPrivateDNSName(id)
+	}
+
+	logrus.Infof("Calling ec2:DescribeInstances for the InstanceId = %s ", id)
+	// Look up instance from EC2 API
+	output, err := p.ec2.DescribeInstances(&ec2.DescribeInstancesInput{
+		InstanceIds: aws.StringSlice([]string{id}),
+	})
+	if err != nil {
+		p.unsetRequestInQueueForInstanceId(id)
+		return "", fmt.Errorf("failed querying private DNS from EC2 API for node %s: %s ", id, err.Error())
+	}
+	for _, reservation := range output.Reservations {
+		for _, instance := range reservation.Instances {
+			if aws.StringValue(instance.InstanceId) == id {
+				privateDNSName = aws.StringValue(instance.PrivateDnsName)
+				p.setPrivateDNSNameCache(id, privateDNSName)
+				p.unsetRequestInQueueForInstanceId(id)
+			}
+		}
+	}
+
+	if privateDNSName == "" {
+		return "", fmt.Errorf("failed to find node %s ", id)
+	}
+	return privateDNSName, nil
+}
+
+func (p *ec2ProviderImpl) StartEc2DescribeBatchProcessing() {
+	startTime := time.Now()
+	var instanceIdList []string
+	for {
+		var instanceId string
+		select {
+		case instanceId = <-p.instanceIdsChannel:
+			logrus.Infof("Received the Instance Id := %s from buffered Channel for batch processing ", instanceId)
```

Comment:
Do you need to log this since you have a log call in the batch process? Can this be debug?

Reply:
Added as Debug log